### ISSUE TITLE
Is a FeatureCollection a set or a sequence?

### ISSUE THREAD

**Issue by dret** [geojson/draft-geojson#147]
https://tools.ietf.org/html/draft-ietf-geojson-00#section-2.3 is very short and says that a FeatureCollection is an array. It does not say whether this array represents a set or a sequence, i.e. whether GeoJSON implementations need to be careful to maintain order when processing GeoJSON. maybe it would be helpful to be explicit about that?

**Comment by adoyle** [comment]
I’d say it’s neither. Existence of duplicate entries or ordering is not addressed by the spec.

By not saying anything I’d say that we’ve said that.

**Comment by calvinmetcalf** [comment]
in context of json doesn't array imply that ordering is important and duplicates are allowed?

**Comment by dret** [comment]
On 2016-02-16 15:43, Allan Doyle wrote:
> I’d say it’s neither. Existence of duplicate entries or ordering is not
> addressed by the spec.
> By not saying anything I’d say that we’ve said that.

i guess technically speaking, that's true. but sometimes it can be 
helpful to spell these things out a bit more clearly.

some people reading the spec might wonder if sequence matters or not. if 
you say it doesn't, then maybe the spec should clearly say that, so that 
nobody starts depending on it and then gets in trouble if an 
implementation does not preserve sequence.

**Comment by dret** [comment]
On 2016-02-16 16:41, Calvin Metcalf wrote:
> in context of json doesn't array imply that ordering is important and
> duplicates are allowed?

that's what JSON would allow you to do, but the question is what GeoJSON 
implementations are allowed to do.

for example, when it comes to ordering object members, then relying on 
sequence would be a bad idea, because of this:

"JSON parsing libraries have been observed to differ as to whether or 
not they make the ordering of object members visible to calling 
software.  Implementations whose behavior does not depend on member 
ordering will be interoperable in the sense that they will not be 
affected by these differences." 
https://tools.ietf.org/html/rfc7159#section-4

that's not the case for arrays, but the question is about the 
significance on the GeoJSON level, not the JSON level.

given that @adoyle said that sequence does not matter, we already have 
two differing interpretations. this might be an indication that 
clarification would be good.

**Comment by adoyle** [comment]
This comes down to the processing model argument. If I have a piece of GeoJSON, the ordering is what it is. What a processing element does to it is out of scope for this document.

**Comment by dret** [comment]
On 2016-02-17 16:29 , Allan Doyle wrote:
> This comes down to the processing model argument. If I have a piece of
> GeoJSON, the ordering is what it is. What a processing element does to
> it is out of scope for this document.

that's equivalent to saying "ordering is important and thus must be 
maintained." in that case a FeatureCollection is a sequence.

**Comment by adoyle** [comment]
That's not what I said.

**Comment by dret** [comment]
On 2016-02-17 16:35 , Allan Doyle wrote:
> That's not what I said.

my apologies then. could you rephrase please and help me understand?

the question is rather simple: if i am a developer writing a piece of 
code reading/writing GeoJSON, i have two slightly conflicting views:

- the FeatureCollection is called a collection which may imply that 
ordering of the features is not relevant.

- the JSON array that is serializing the features has sequence semantics 
on the JSON level, nobody feels free to simply re-arrange an array 
(unless being told otherwise).

the spec should tell the dev what to do. does he have to go through the 
effort and maintain sequence when parsing/re-serializing a 
FeatureCollection? if he doesn't do it, will stuff break?

this issue simply tries to take the guesswork out of this and add pretty 
much sentence telling the dev with the above question what to do.

**Comment by adoyle** [comment]
OK. I decided to look it up. From what I can tell, FeatureCollections in the context of a WFS are ordered. They are a response to a query and the query has an optional "order by" directive.

http://docs.opengeospatial.org/is/09-025r2/09-025r2.html

**Comment by dret** [comment]
@adoyle, so then it would be fine to call it a sequence, from your point of view?

**Comment by sgillies** [comment]
@dret originally, the spec said a feature collection contained an array of features. JSON has no concept of a set, and we didn't add our own. I think we should continue with "features" as a sequence, but call it a "list" to distinguish it from the GeoJSON text sequences we're going to work on elsewhere.

**Comment by dret** [comment]
On 2016-02-17 23:21, Sean Gillies wrote:
> @dret <https://github.com/dret> originally, the spec said a feature
> collection contained an array of features. JSON has no concept of a set,
> and we didn't add our own. I think we should continue with "features" as
> a sequence, but call it a "list" to distinguish it from the GeoJSON text
> sequences we're going to work on elsewhere.

i am fine with anything that makes it clear to devs that they cannot 
parse a feature collection, and then reserialize it, without making sure 
that order is maintained. that's probably a single sentence that we can 
add to make that crystal clear.

**Comment by mpdaly** [comment]
https://www.rfc-editor.org/rfc/rfc7159.txt says "An array is an ordered sequence of zero or more values."

There are no requirements to maintain the order articulated in that RFC, so I see no reason to articulate them in this one.

**Comment by sdrees** [comment]
I second @mpdaly to not want to add to the language. I ammend that adding to this would also put an adittional burden on the producer to maintain semantics or at least a stable ordering when constructing the array which than should have been augmented into a named element of the GeoJSON specification in the first place. So I see more questions arising in processing on both sides of the interchange format - producer and consumer - we might not really want to answer in this decade ...

**Comment by calvinmetcalf** [comment]
I would also be against making order mandatory to be preserved, while order
is significant I see no real benefit of mandating it be preserved.

**Comment by dret** [comment]
it seems that there's a majority for not explicitly saying whether a collection should be considered a set or a sequence. i am suggesting to leave this open for final comments if anybody feels like it, and then to close it with no action.
